File: src/shared/units.ts

import unitsData from "./units.json";

export type UnitTier = "citizen" | "guard" | "archer" | "warrior" | "knight";
export type DamageRangeType = "melee" | "ranged";

export interface UnitCost {
    wood: number;
    stone: number;
    iron: number;
}

export interface BlockProfile {
    meleeDR: number;
    rangedDR: number;
    drainPerSecond: number;
    recoverySeconds?: number;
}

export interface PositionalBonus {
    staminaCostReductionPct: number;
    minNearbyAllies: number;
}

export interface OpportunistBow {
    damage: number;
    range: number;
    cooldownSeconds: number;
}

export interface RogueQuiver {
    damage: number;
    range: number;
    charges: number;
    rechargeSeconds: number;
    executeMultiplier: number;
    executeHpThresholdPct: number;
}

export interface Recon {
    viewRadiusBonusPct: number;
    detectionRadiusMeters: number;
}

export interface FireArrow {
    damage: number;
    structureDamageMultiplier: number;
    woodCostPerShot: number;
}

export interface DashThrust {
    distanceMeters: number;
    windupSeconds: number;
    recoverySeconds: number;
    damageMultiplier: number;
    cooldownSeconds: number;
}

export interface UnitDefinition {
    /** Stable numeric id used on the wire (WELCOME / unit-roster messages). */
    typeId: number;
    /** Stable string key used in code/config/URLs. */
    id: string;
    displayName: string;
    tier: UnitTier;

    hp: number;
    /** Passive damage reduction (0..1), always active. Only Heavy Knight/Paladin have this. */
    passiveDR: number;
    /** m/s */
    moveSpeed: number;
    rangeType: DamageRangeType;
    /** meters */
    range: number;
    damage: number;
    windupSeconds: number;
    activeSeconds: number;
    recoverySeconds: number;

    stamina: number;
    staminaRegenPerSecond: number;
    /** Sprint (GDD §54): move-speed multiplier while holding Shift and moving. */
    sprintSpeedMultiplier: number;
    /** Sprint (GDD §57): stamina drained per second while sprinting. */
    sprintStaminaCostPerSecond: number;
    /** AnimatedSprite.animationSpeed (frames/tick) for every animation this unit plays. */
    animationSpeed: number;
    /** Attack combo chain length (undefined/1 = no combo, every swing re-attacks). */
    comboSteps?: number;
    /** How long past a swing's own duration a new swing still continues the combo
     * chain instead of resetting to step 1 — see server world.go executeAttack. */
    comboWindowSeconds?: number;
    /** Flat stamina cost per swing/thrust, where the doc specifies one. */
    attackStaminaCost?: number;
    /** Archer-only: draw-and-hold only costs stamina past this many seconds. */
    drawHoldThresholdSeconds?: number;
    dodgeCostMultiplier?: number;

    cost: UnitCost;
    requiresRoyalGuard?: boolean;
    cleave?: boolean;
    hasBraceStance?: boolean;

    block?: BlockProfile;
    positionalBonus?: PositionalBonus;
    opportunistBow?: OpportunistBow;
    rogueQuiver?: RogueQuiver;
    recon?: Recon;
    fireArrow?: FireArrow;
    dashThrust?: DashThrust;
    antiShieldMultiplier?: number;
    antiWoodStructureMultiplier?: number;

    /** Reference to public/assets/<path> — not yet wired into the sprite loader. */
    assetPath?: string;
    combatAssetPath?: string;
    dashAssetPath?: string;
}

export const UNITS: readonly UnitDefinition[] = unitsData as UnitDefinition[];

const BY_ID = new Map<string, UnitDefinition>(UNITS.map((u) => [u.id, u]));
const BY_TYPE_ID = new Map<number, UnitDefinition>(UNITS.map((u) => [u.typeId, u]));

export type UnitType = (typeof UNITS)[number]["id"];

export const DEFAULT_UNIT_TYPE: UnitType = "spearman";

export function isValidUnitType(id: string): id is UnitType {
    return BY_ID.has(id);
}

export function getUnitDefinition(id: string): UnitDefinition {
    return BY_ID.get(id) ?? BY_ID.get(DEFAULT_UNIT_TYPE)!;
}

export function getUnitDefinitionByTypeId(typeId: number): UnitDefinition {
    return BY_TYPE_ID.get(typeId) ?? BY_ID.get(DEFAULT_UNIT_TYPE)!;
}

File: src/server/internal/units/units.go
package units

import (
	_ "embed"
	"encoding/json"
	"fmt"
)

//go:embed units.json
var embeddedUnits []byte

type Cost struct {
	Wood  int `json:"wood"`
	Stone int `json:"stone"`
	Iron  int `json:"iron"`
}

type BlockProfile struct {
	MeleeDR         float64  `json:"meleeDR"`
	RangedDR        float64  `json:"rangedDR"`
	DrainPerSecond  float64  `json:"drainPerSecond"`
	RecoverySeconds *float64 `json:"recoverySeconds,omitempty"`
}

type PositionalBonus struct {
	StaminaCostReductionPct float64 `json:"staminaCostReductionPct"`
	MinNearbyAllies         int     `json:"minNearbyAllies"`
}

type OpportunistBow struct {
	Damage          float64 `json:"damage"`
	Range           float64 `json:"range"`
	CooldownSeconds float64 `json:"cooldownSeconds"`
}

type RogueQuiver struct {
	Damage                float64 `json:"damage"`
	Range                 float64 `json:"range"`
	Charges               int     `json:"charges"`
	RechargeSeconds       float64 `json:"rechargeSeconds"`
	ExecuteMultiplier     float64 `json:"executeMultiplier"`
	ExecuteHpThresholdPct float64 `json:"executeHpThresholdPct"`
}

type Recon struct {
	ViewRadiusBonusPct    float64 `json:"viewRadiusBonusPct"`
	DetectionRadiusMeters float64 `json:"detectionRadiusMeters"`
}

type FireArrow struct {
	Damage                    float64 `json:"damage"`
	StructureDamageMultiplier float64 `json:"structureDamageMultiplier"`
	WoodCostPerShot           int     `json:"woodCostPerShot"`
}

type DashThrust struct {
	DistanceMeters   float64 `json:"distanceMeters"`
	WindupSeconds    float64 `json:"windupSeconds"`
	RecoverySeconds  float64 `json:"recoverySeconds"`
	DamageMultiplier float64 `json:"damageMultiplier"`
	CooldownSeconds  float64 `json:"cooldownSeconds"`
}

type Definition struct {
	TypeID      uint8  `json:"typeId"`
	ID          string `json:"id"`
	DisplayName string `json:"displayName"`
	Tier        string `json:"tier"`

	HP              float64 `json:"hp"`
	PassiveDR       float64 `json:"passiveDR"`
	MoveSpeed       float64 `json:"moveSpeed"`
	RangeType       string  `json:"rangeType"`
	Range           float64 `json:"range"`
	Damage          float64 `json:"damage"`
	WindupSeconds   float64 `json:"windupSeconds"`
	ActiveSeconds   float64 `json:"activeSeconds"`
	RecoverySeconds float64 `json:"recoverySeconds"`

	Stamina                    float64 `json:"stamina"`
	StaminaRegenPerSecond      float64 `json:"staminaRegenPerSecond"`
	SprintSpeedMultiplier      float64 `json:"sprintSpeedMultiplier"`
	SprintStaminaCostPerSecond float64 `json:"sprintStaminaCostPerSecond"`

	ComboSteps         int     `json:"comboSteps,omitempty"`
	ComboWindowSeconds float64 `json:"comboWindowSeconds,omitempty"`

	AnimationSpeed           float64  `json:"animationSpeed"`
	AttackStaminaCost        *float64 `json:"attackStaminaCost,omitempty"`
	DrawHoldThresholdSeconds *float64 `json:"drawHoldThresholdSeconds,omitempty"`
	DodgeCostMultiplier      *float64 `json:"dodgeCostMultiplier,omitempty"`

	Cost               Cost `json:"cost"`
	RequiresRoyalGuard bool `json:"requiresRoyalGuard,omitempty"`
	Cleave             bool `json:"cleave,omitempty"`
	HasBraceStance     bool `json:"hasBraceStance,omitempty"`

	Block                       *BlockProfile    `json:"block,omitempty"`
	PositionalBonus             *PositionalBonus `json:"positionalBonus,omitempty"`
	OpportunistBow              *OpportunistBow  `json:"opportunistBow,omitempty"`
	RogueQuiver                 *RogueQuiver     `json:"rogueQuiver,omitempty"`
	Recon                       *Recon           `json:"recon,omitempty"`
	FireArrow                   *FireArrow       `json:"fireArrow,omitempty"`
	DashThrust                  *DashThrust      `json:"dashThrust,omitempty"`
	AntiShieldMultiplier        *float64         `json:"antiShieldMultiplier,omitempty"`
	AntiWoodStructureMultiplier *float64         `json:"antiWoodStructureMultiplier,omitempty"`

	AssetPath       string `json:"assetPath,omitempty"`
	CombatAssetPath string `json:"combatAssetPath,omitempty"`
	DashAssetPath   string `json:"dashAssetPath,omitempty"`
}

var (
	all      []Definition
	byID     map[string]Definition
	byTypeID map[uint8]Definition
)

const DefaultUnitType = "spearman"

func init() {
	if err := json.Unmarshal(embeddedUnits, &all); err != nil {
		panic(fmt.Errorf("units: failed to parse embedded units.json: %w", err))
	}
	byID = make(map[string]Definition, len(all))
	byTypeID = make(map[uint8]Definition, len(all))
	for _, u := range all {
		byID[u.ID] = u
		byTypeID[u.TypeID] = u
	}
	if _, ok := byID[DefaultUnitType]; !ok {
		panic(fmt.Errorf("units: DefaultUnitType %q not found in units.json", DefaultUnitType))
	}
}

func All() []Definition {
	return all
}

func Get(id string) Definition {
	if u, ok := byID[id]; ok {
		return u
	}
	return byID[DefaultUnitType]
}

func GetByTypeID(typeID uint8) Definition {
	if u, ok := byTypeID[typeID]; ok {
		return u
	}
	return byID[DefaultUnitType]
}

func IsValid(id string) bool {
	_, ok := byID[id]
	return ok
}
